### PR TITLE
New Array methods

### PR DESCRIPTION
**Pull request by mrvoorhe**

### CODE REVIEW

File: unity/unity-embed-host/CoreCLRHost.cs [hunk 1]
```diff
@@ -99,6 +99,18 @@ public static IntPtr object_isinst([NativeCallbackType("MonoObject*")] IntPtr ob
     public static IntPtr object_get_class([NativeCallbackType("MonoObject*")] IntPtr obj)
         => obj.ToManagedRepresentation().TypeHandleIntPtr();
 
+    [return: NativeCallbackType("MonoObject*")]
+    public static IntPtr array_new([NativeCallbackType("MonoObject*")] IntPtr domain, [NativeCallbackType("MonoClass*")] IntPtr klass, [NativeWrapperType("guint32")] int n)
+        => Array.CreateInstance(klass.TypeFromHandleIntPtr(), n).ToNativeRepresentation();
```
Comment (joncham) [inline unity/unity-embed-host/CoreCLRHost.cs:104]: This is where performance may start to come into play. I think it's fine to proceed with correctness first.

Questions worth considering
1. How does this implementation perform compared to the previous C++ code? 
2. To what extent does Unity use this API and what performance expectations does it have?
3. What are the options to improve this performance? Maybe a `Dictionary<IntPtr, Func<object>>` that would be type specific functions to allocate an object of a given type.

**Comment by joncham** [comment]
Are you doing any testing in Unity of the changes being made to the embedding API implementation? It may be worth running locally if nothing else, to verify any large changes don't cause failures to CoreCLR runtime tests.

File: unity/unity-embed-host/CoreCLRHost.cs (hunk 1 above)
Comment (mrvoorhe) [reply unity/unity-embed-host/CoreCLRHost.cs:104]: As I add apis, I think it would make more sense to handle performance review separately.   At least when a trivial implementation is possible.  The advantages being

*  By getting an implementation in, we unblock other efforts such as if/how the testing approach will evolve.
*  Performance testing will require additional setup.  W/e that setup is, I think it would be nice to have a few apis to look at rather than look at 1 and then switch off of performance.
*  Depending on how the testing approach evolves, maybe that would impact the approach we take to performance testing.

I made https://jira.unity3d.com/browse/VM-1974 to start capturing the apis we should come back to.

**Comment by mrvoorhe** [comment]
> Are you doing any testing in Unity of the changes being made to the embedding API implementation? It may be worth running locally if nothing else, to verify any large changes don't cause failures to CoreCLR runtime tests.

I have not been.  I was under the impression that if something had a native test and was straightforward to implement then the change should be OK on the Unity side.  Am I placing too much faith in the tests that exist in this repo?

**Comment by UnityAlex** [comment]
> > Are you doing any testing in Unity of the changes being made to the embedding API implementation? It may be worth running locally if nothing else, to verify any large changes don't cause failures to CoreCLR runtime tests.
> 
> I have not been. I was under the impression that if something had a native test and was straightforward to implement then the change should be OK on the Unity side. Am I placing too much faith in the tests that exist in this repo?

I don't think we can put too much faith in the existing tests in this repository as the coreclr runtime hasn't seen a lot of widespread use to catch a lot of the corner casey stuff that would trigger additional tests to be added.

File: unity/unity-embed-host/CoreCLRHost.cs (hunk 1 above)
Comment (joncham) [reply unity/unity-embed-host/CoreCLRHost.cs:104]: Agreed and thanks!

**Comment by joncham** [comment]
> Am I placing too much faith in the tests that exist in this repo?

I think they provide reasonable coverage. My recommendation is to try Unity itself at reasonable intervals, otherwise you may rewrite the entire embedding API and then have a large set of changes to debug if you have issues integrating.

**Comment by mrvoorhe** [comment]
@joncham @UnityAlex Tests are green now.  Please take a look.

**Review by UnityAlex — APPROVED** [review]
Nice and clean!

**Review by joncham — APPROVED** [review]
